lib/rules/no-useless-escape.js

```diff
@@ -182,6 +150,131 @@ module.exports = {
+            let patternNode;
+            const unicode = flags.includes("u");
+            const unicodeSets = flags.includes("v");
+
+            try {
+                patternNode = parser.parsePattern(pattern, 0, pattern.length, { unicode, unicodeSets });
+            } catch {
+
+                // Ignore regular expressions with syntax errors
+                return;
+            }
+
+            /** @type {(CharacterClass | ExpressionCharacterClass)[]} */
+            const characterClassStack = [];
+
+            visitRegExpAST(patternNode, {
+                onCharacterClassEnter: characterClassNode => characterClassStack.unshift(characterClassNode),
+                onCharacterClassLeave: () => characterClassStack.shift(),
+                onExpressionCharacterClassEnter: characterClassNode => characterClassStack.unshift(characterClassNode),
+                onExpressionCharacterClassLeave: () => characterClassStack.shift(),
+                onCharacterEnter(characterNode) {
+                    if (!characterNode.raw.startsWith("\\")) {
+
+                        // It's not an escaped character.
+                        return;
+                    }
+
+                    const escapedChar = characterNode.raw.slice(1);
+
+                    if (escapedChar !== String.fromCodePoint(characterNode.value)) {
+
+                        // It's a valid escape.
+                        return;
+                    }
+                    let allowedEscapes;
+
+                    if (characterClassStack.length) {
+                        allowedEscapes = unicodeSets ? REGEX_CLASSSET_CHARACTER_ESCAPES : REGEX_GENERAL_ESCAPES;
+                    } else {
+                        allowedEscapes = REGEX_NON_CHARCLASS_ESCAPES;
+                    }
+                    if (allowedEscapes.has(escapedChar)) {
+                        return;
+                    }
+
+                    const reportedIndex = characterNode.start + 1;
+
+                    if (characterClassStack.length) {
+                        if (!unicodeSets) {
+                            const characterClassNode = characterClassStack[0];
+
+                            if (characterClassNode.type === "CharacterClass") {
+                                if (escapedChar === "-") {
+
+                                    /*
+                                     * The '-' character is a special case, because it's only valid to escape it if it's in a character
+                                     * class, and is not at either edge of the character class. To account for this, don't consider '-'
+                                     * characters to be valid in general, and filter out '-' characters that appear in the middle of a
+                                     * character class.
+                                     */
+                                    if (characterClassNode.start + 1 !== characterNode.start && characterNode.end !== characterClassNode.end - 1) {
```

Comment:
I think this is a false negative:

```js
/* eslint no-useless-escape: "error" */

/[^\-a]/
```

Reply:
It's a false negative, but I think it's the same with the current rule.
May I fix that issue in this PR?

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZWxlc3MtZXNjYXBlOiBcImVycm9yXCIgKi9cblxuL1teXFwtYV0vIiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6eyJwYWRkaW5nLWxpbmUtYmV0d2Vlbi1zdGF0ZW1lbnRzIjpbImVycm9yIix7ImJsYW5rTGluZSI6ImFsd2F5cyIsInByZXYiOiIqIiwibmV4dCI6IioifV19LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCJ9fX0=

Reply:
Ah okay, since it's existing behavior then seems better to address that separately from this PR.